src/sa.py

```diff
@@ -1,15 +1,277 @@
+
+def bucket_indices(bucket_out):
+    indices = {}
+    for key in bucket_out:
+        indices[key] = len(bucket_out[key]) - 1
+    return indices
+
+def third_pass(bucket_out, LS_array, string):
+    indices = bucket_indices(bucket_out)
+
+    for key in sorted(bucket_out, reverse=True):
+        for i in range(len(bucket_out[key]) - 1, -1, -1):
+            if bucket_out[key][i] < 1:
+                continue
+            elif LS_array[bucket_out[key][i] - 1] == "S":
+                buck = string[bucket_out[key][i] - 1]
+                bucket_out[buck][indices[buck]] = bucket_out[key][i] - 1
+                indices[buck] -= 1
+    return bucket_out
+
+def number_LMS(bucket_out, LMS_blocks, LMS_array):
+    reduced = [0 for i in range(len(LMS_array))]
+    previous = None
+    count = -1
+    for key in bucket_out:
+        for element in bucket_out[key]:
+            if element in LMS_array:
+                block = LMS_blocks[LMS_array.index(element)]
+                if block != previous:
+                    count += 1
+                previous = block
+                reduced[LMS_array.index(element)] = str(count)
+    return "".join(reduced)
+
+def compute_t1(string):
+    LS_array = annotate_string(string)
+    LMS_array = get_LMS(LS_array)
+    LMS_blocks = get_LMS_blocks(string, LMS_array)
+    buckets = impute_buckets(string)
+    bucket_out = output_bucket(buckets)
+    bucket_out = first_pass(buckets, bucket_out, LMS_array, LMS_blocks)
+    bucket_out = second_pass(bucket_out, LS_array, string)
+    bucket_out = third_pass(bucket_out, LS_array, string)
+    reduced = 0#number_LMS(bucket_out, LMS_blocks, LMS_array)
+    return reduced, bucket_out
+
+#def SAIS(string):
+#    stack = list()
+#    while True:
+#        reduced, bucket_out = compute_t1
+#        if reduced[-1] != len()
+
+#def mostly_naive(string):
+#    forget, bucket_out = compute_t1(string)
+#    SA = []
+#    for i in bucket_out:
+#        SA.extend(bucket_out[i])
+#    SA = [string[SA[int(i)]:] for i in SA]
+#    SA.sort()
+#    return(SA)
+
+def completely_naive(string):
+    SA = [i for i in range(len(string))]
+    SA = [string[SA[int(i)]:] for i in SA]
+    SA.sort()
+    return(SA)
+
+def test_SA(SA):
+    for i in SA:
+        print(i)
+
+def LCP_naive(SA, string):
+    LCP = [0, 0]
+    for i in range(2, len(SA)):
+        count = 0
+        for j in range(len(string)):
+            if string[SA[i] + j] == string[SA[i - 1] + j]:
+                count += 1
+            else:
+                break
+        LCP.append(count)
+    return LCP
+
+def naive(string):
+    string += "$"
+    SA = completely_naive(string)
+    SA = [len(string) - len(i) for i in SA]
+    LCP = LCP_naive(SA, string)
+    return SA, LCP
+
+def compare_strings(pattern, suffix):
+    loops = min(len(pattern), len(suffix))
+
+    for i in range(loops):
+        if pattern[i] < suffix[i]:
+            return False
+        if pattern[i] > suffix[i]:
+            return True

```

Comment:
Comparing two strings by iterating through them in Python will be slow. The builtin comparison runs in C and is faster. So you could ask

```python
if pattern < suffix:
   return False
if pattern > suffix:
   return True
return None
```

(I find the return values confusing, but I kept them for this)